File: framework/pure_nn/module.py
from .layers import Layer
from .losses import Loss


class Module:
    def __init__(self):
        self._parameters = {}
        self._layers = []

    def __call__(self, x):
        return self.forward(x)

    def forward(self, x):
        raise NotImplementedError

    @property
    def parameters(self):
        return self._parameters

    def __setattr__(self, key, value):
        if isinstance(value, Layer):
            layer = value
            self._parameters[key] = layer.vars
            self._layers.append(value)
        object.__setattr__(self, key, value)

    def backward(self, loss):
        assert isinstance(loss, Loss)
        delta = loss.delta
        for layer in self._layers[::-1]:
            delta = layer.backward(delta)

File: cpp_nn/module.cpp
#include "module.h"

void Module::backward(const Loss &loss)
{
    float_batch delta = loss.delta;
//    cout<<delta[0][0]<<endl;
    size_t num_layers = this->parameters.size();
    for(int i = num_layers - 1; i >= 0; i--){
        delta = this->parameters[i]->backward(delta);
    }

}
